src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/SolutionExplorerInProcess.cs

```diff
@@ -682,4 +709,160 @@ private async Task<EnvDTE.Project> GetProjectAsync(string nameOrFileName, Cancel
                 });
         }
     }
+
+    internal sealed class UpdateSolutionEvents : IVsUpdateSolutionEvents, IVsUpdateSolutionEvents2, IDisposable
+    {
+        private uint _cookie;
+        private readonly IVsSolutionBuildManager2 _solutionBuildManager;
+
+        internal delegate void UpdateSolutionDoneEvent(bool succeeded, bool modified, bool canceled);
+
+        internal delegate void UpdateSolutionBeginEvent(ref bool cancel);
+
+        internal delegate void UpdateSolutionStartUpdateEvent(ref bool cancel);
+
+        internal delegate void UpdateProjectConfigDoneEvent(IVsHierarchy projectHierarchy, IVsCfg projectConfig, int success);
+
+        internal delegate void UpdateProjectConfigBeginEvent(IVsHierarchy projectHierarchy, IVsCfg projectConfig);
+
+        public event UpdateSolutionDoneEvent? OnUpdateSolutionDone;
+
+        public event UpdateSolutionBeginEvent? OnUpdateSolutionBegin;
+
+        public event UpdateSolutionStartUpdateEvent? OnUpdateSolutionStartUpdate;
+
+        public event Action? OnActiveProjectConfigurationChange;
+
+        public event Action? OnUpdateSolutionCancel;
+
+        public event UpdateProjectConfigDoneEvent? OnUpdateProjectConfigDone;
+
+        public event UpdateProjectConfigBeginEvent? OnUpdateProjectConfigBegin;
+
+        internal UpdateSolutionEvents(IVsSolutionBuildManager2 solutionBuildManager)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            _solutionBuildManager = solutionBuildManager;
+            ErrorHandler.ThrowOnFailure(solutionBuildManager.AdviseUpdateSolutionEvents(this, out _cookie));
+        }
+
+        int IVsUpdateSolutionEvents.UpdateSolution_Begin(ref int pfCancelUpdate)
+        {
+            var cancel = false;
+            OnUpdateSolutionBegin?.Invoke(ref cancel);
+            if (cancel)
+            {
+                pfCancelUpdate = 1;
+            }
+
+            return 0;
+        }
+
+        int IVsUpdateSolutionEvents.UpdateSolution_Done(int fSucceeded, int fModified, int fCancelCommand)
+        {
+            OnUpdateSolutionDone?.Invoke(fSucceeded != 0, fModified != 0, fCancelCommand != 0);
```

Comment:
it looks like we just need IVsUpdateSolutionEvents.UpdateSolution_Done.  Is there a reason we need to implement IVsUpdateSolutionEvents2 as well?

Reply:
Let me admit I took this whole hog from https://github.com/dotnet/roslyn-sdk/pull/532. I will clean it up.

Reply:
Updated to remove unused code